File: DSA/Manav-Rajput/Day11_question_1.java

import java.util.LinkedList;
import java.util.Queue;

class Main {
    static class Cell {
        int row, col, dist;
        
        Cell(int row, int col, int dist) {
            this.row = row;
            this.col = col;
            this.dist = dist;
        }
    }

    public static int minMoves(char[][] grid, int[] start, int[] goal) {
        int rows = grid.length;
        int cols = grid[0].length;

        boolean[][] visited = new boolean[rows][cols];
        
        // Directions for moving up, down, left, right
        int[] rowDirection = {-1, 1, 0, 0};
        int[] colDirection = {0, 0, -1, 1};

        Queue<Cell> queue = new LinkedList<>();
        queue.add(new Cell(start[0], start[1], 0));
        visited[start[0]][start[1]] = true;

        while (!queue.isEmpty()) {
            Cell current = queue.poll();

            // If we reached the goal
            if (current.row == goal[0] && current.col == goal[1]) {
                return current.dist;
            }

            // Explore neighbors
            for (int i = 0; i < 4; i++) {
                int newRow = current.row + rowDirection[i];
                int newCol = current.col + colDirection[i];

                // Check if the new position is within bounds and not visited or blocked
                if (newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols &&
                    !visited[newRow][newCol] && grid[newRow][newCol] == '.') {
                    queue.add(new Cell(newRow, newCol, current.dist + 1));
                    visited[newRow][newCol] = true;
                }
            }
        }

        // If we can't reach the goal
        return -1;
    }

    public static void main(String[] args) {
        char[][] grid = {
            {'.', '.', '.'},
            {'.', 'X', '.'},
            {'.', '.', '.'}
        };
        int[] start = {0, 0};
        int[] goal = {2, 2};

        int result = minMoves(grid, start, goal);
        System.out.println("Minimum moves required: " + result);
    }
}

File: DSA/Arshdeep-Kaur/Day11_Question1.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

const vector<pair<int, int>> directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

bool isValid(int x, int y, int rows, int cols, vector<vector<char>>& grid) {
    return x >= 0 && x < rows && y >= 0 && y < cols && grid[x][y] == '.';
}

int minMoves(vector<vector<char>>& grid, pair<int, int> start, pair<int, int> goal) {
    int rows = grid.size(), cols = grid[0].size();
    queue<pair<int, int>> q;
    q.push(start);
    grid[start.first][start.second] = 'X';
    int moves = 0;

    while (!q.empty()) {
        int size = q.size();
        for (int i = 0; i < size; ++i) {
            auto [x, y] = q.front(); q.pop();
            if (x == goal.first && y == goal.second) {
                return moves;
            }

            for (auto [dx, dy] : directions) {
                int newX = x + dx, newY = y + dy;
                if (isValid(newX, newY, rows, cols, grid)) {
                    q.push({newX, newY});
                    grid[newX][newY] = 'X';
                }
            }
        }
        ++moves;
    }

    return -1; // Goal not reachable
}

int main() {
    vector<vector<char>> grid = {
        {'.', '.', '.'},
        {'.', 'X', '.'},
        {'.', '.', '.'}
    };
    pair<int, int> start = {0, 0};
    pair<int, int> goal = {2, 2};

    int result = minMoves(grid, start, goal);

    if (result != -1) {
        cout << result << endl;
    } else {
        cout << "No path found from start to goal." << endl;
    }

    return 0;
}
